src/core/net/dnssd_server.cpp

```diff
@@ -839,6 +899,62 @@ Error Server::ResolveByQueryCallbacks(Header                 &aResponseHeader,
     return error;
 }
 
+#if OPENTHREAD_CONFIG_DNS_UPSTREAM_QUERY_ENABLE
+void Server::OnUpstreamQueryResponse(UpstreamQueryTransaction &aQueryTransaction, Message &aResponseMessage)
+{
+    Error err = mSocket.SendTo(aResponseMessage, aQueryTransaction.mMessageInfo);
+
+    if (err != kErrorNone)
+    {
+        aResponseMessage.Free();
+    }
+
+    aQueryTransaction.Reset();
+
+    ResetTimer();
+}
+
+Server::UpstreamQueryTransaction *Server::AllocateUpstreamQueryTransaction(const Ip6::MessageInfo &aMessageInfo)
+{
+    UpstreamQueryTransaction *ret = nullptr;
+
+    for (UpstreamQueryTransaction &txn : mUpstreamQueryTransactions)
+    {
+        if (!txn.IsValid())
+        {
+            ret              = &txn;
+            txn.mValid       = true;
+            txn.mMessageInfo = aMessageInfo;
+            txn.mStartTime   = TimerMilli::GetNow();
+            break;
+        }
+    }
+
+    if (ret != nullptr)
+    {
+        ResetTimer();
```

Comment:
Can do `mTimer.FireAtIfEarlier(ret->GetExpireTime())` instead?
(since we only add a new txn with new timeout)